src/passes/Precompute.cpp

```diff
@@ -228,6 +223,12 @@ struct Precompute
   // Precompute an expression, returning a flow, which may be a constant
   // (that we can replace the expression with if replaceExpression is set).
   Flow precomputeExpression(Expression* curr, bool replaceExpression = true) {
+    // Don't try to precompute a reference. We can't replace it with a constant
+    // expression, as that would make a copy of it by value.
+    // TODO: do so when safe
+    if (curr->type.isRef()) {
+      return Flow(NONCONSTANT_FLOW);
+    }
```

Comment:
I don't quite understand why moving this makes a difference. What's a case where the previous code would have been incorrect? Would moving it to the top of `visitExpression` work as well? If so, that seems like a better place for this.

Reply:
The previous location was in `visitExpression` right before a call to `precomputeExpression`. The new location is in `precomputeExpression` itself, which guarantees that all callers have the right behavior (which includes `precomputeValue` in particular, where the new bug was).

Reply:
Got it, thanks!